src/generic/linalg.jl

```diff
@@ -0,0 +1,478 @@
+    end
+    if Typ1 ≠ Typ2
+        ret = quote
+            $ret
+            BandedMatrices.@_banded_banded_linalg($Typ2, $Typ1)
+            BandedMatrices.@banded_banded_banded_linalg($Typ1, $Typ1, $Typ2)
+            BandedMatrices.@banded_banded_banded_linalg($Typ2, $Typ2, $Typ1)
+        end
+    end
+    esc(ret)
+end
+
+# add banded linear algebra routines for Typ implementing the BandedMatrix interface
+macro banded_linalg(Typ)
+    ret = quote
+        BandedMatrices.@banded_banded_linalg($Typ, $Typ)
+        # default is to use dense axpy!
+        BandedMatrices.banded_axpy!(a::Number, X::$Typ, Y::AbstractMatrix) =
+                BandedMatrices.banded_dense_axpy!(a, X, Y)
+        Base.BLAS.axpy!(a::Number, X::$Typ, Y::AbstractMatrix) =
+                BandedMatrices.banded_axpy!(a, X, Y)
+        function Base.:+(A::$Typ{T}, B::AbstractMatrix{T}) where {T}
+            ret = deepcopy(B)
+            axpy!(one(T), A, ret)
+            ret
+        end
+        function Base.:+(A::$Typ{T}, B::AbstractMatrix{V}) where {T,V}
+            n, m=size(A)
+            ret = zeros(promote_type(T,V),n,m)
+            axpy!(one(T), A,ret)
+            axpy!(one(V), B,ret)
+            ret
+        end
+        Base.:+(A::AbstractMatrix{T}, B::$Typ{V}) where {T,V} = B + A
+
+        function Base.:-(A::$Typ{T}, B::AbstractMatrix{T}) where {T}
+            ret = deepcopy(B)
+            Base.scale!(ret,-1)
+            Base.BLAS.axpy!(one(T), A, ret)
+            ret
+        end
+        function Base.:-(A::$Typ{T}, B::AbstractMatrix{V}) where {T,V}
+            n, m= size(A)
+            ret = zeros(promote_type(T,V),n,m)
+            Base.BLAS.axpy!(one(T),  A, ret)
+            Base.BLAS.axpy!(-one(V), B, ret)
+            ret
+        end
+        Base.:-(A::AbstractMatrix{T}, B::$Typ{V}) where {T,V} = Base.scale!(B - A, -1)
+
+
+
+        ## UniformScaling
+
+        function Base.BLAS.axpy!(a::Number, X::UniformScaling, Y::$Typ{T}) where {T}
+            BandedMatrices.checksquare(Y)
+            α = a * X.λ
+            @inbounds for k = 1:size(Y,1)
+                BandedMatrices.inbands_setindex!(Y, BandedMatrices.inbands_getindex(Y, k, k) + α, k, k)
+            end
+            Y
+        end
+
+        function Base.:+(A::$Typ, B::UniformScaling)
+            ret = deepcopy(A)
+            Base.BLAS.axpy!(1, B,ret)
+        end
+
+        Base.:+(A::UniformScaling, B::$Typ) = B + A
+
+        function Base.:-(A::$Typ, B::UniformScaling)
+            ret = deepcopy(A)
+            axpy!(-1, B, ret)
+        end
+
+        function Base.:-(A::UniformScaling, B::$Typ)
+            ret = deepcopy(B)
+            Base.scale!(ret, -1)
+            axpy!(1, A, ret)
+        end
+
+        # use BLAS routine for positively banded BLASBandedMatrix
+        BandedMatrices.positively_banded_matvecmul!(c::StridedVector{T}, tA::Char, A::$Typ{T}, b::StridedVector{T}) where {T <: BandedMatrices.BlasFloat} =
```

Comment:
this function should be only defined when the interface `leadingdimension(::Typ{T})` is defined.